File: src/test/java/nihilite/test/redefineInstanceDriver.java
package nihilite.test;

import clojure.java.api.Clojure;
import clojure.lang.IFn;
import net.bytebuddy.agent.ByteBuddyAgent;
import java.lang.instrument.Instrumentation;
import java.util.concurrent.atomic.AtomicReference;

public class redefineInstanceDriver {

    public static volatile AtomicReference<Object> CAPTURED = new AtomicReference<>();

    public int counter = 0;

    public String probe() {
        return "ORIGINAL-COUNTER=" + counter;
    }

    public static void main(String[] args) throws Exception {
        Instrumentation inst = ByteBuddyAgent.install();
        nihilite.agent.Agent.agentmain(null, inst);
        nihilite.hooks.HookInstaller.install(inst);

        IFn require = Clojure.var("clojure.core", "require");
        require.invoke(Clojure.read("nihilite.registry"));
        require.invoke(Clojure.read("nihilite.api"));

        IFn installHook = Clojure.var("nihilite.registry", "install!");
        IFn clearHook = Clojure.var("nihilite.registry", "clear!");

        IFn bridge = new clojure.lang.AFn() {
            @Override public Object invoke(Object self, Object args, Object methodName) {
                CAPTURED.set(self);
                return "REDEFINED-AT-ARITY0";
            }
        };

        IFn hashMap = Clojure.var("clojure.core", "hash-map");
        installHook.invoke(hashMap.invoke(
                Clojure.read(":id"),                 "inst-redef",
                Clojure.read(":target-internal"),    "nihilite/test/redefineInstanceDriver",
                Clojure.read(":method-name"),        "probe",
                Clojure.read(":position"),           Clojure.read(":redefine"),
                Clojure.read(":arity"),              0,
                Clojure.read(":descriptor"),         "()Ljava/lang/String;",
                Clojure.read(":bridge"),             bridge));

        IFn installRedisp = Clojure.var("nihilite.registry", "install-redefine-dispatcher!");
        installRedisp.invoke();

        redefineInstanceDriver instance = new redefineInstanceDriver();
        instance.counter = 99;
        String result = instance.probe();

        if (!"REDEFINED-AT-ARITY0".equals(result)) {
            System.err.println("redefineInstanceDriver FAIL: probe was \"" + result
                    + "\" expected \"REDEFINED-AT-ARITY0\"");
            System.exit(3);
        }
        Object captured = CAPTURED.get();
        if (captured == null) {
            System.err.println("redefineInstanceDriver FAIL: bridge did not run (CAPTURED null)");
            System.exit(4);
        }
        if (captured != instance) {
            System.err.println("redefineInstanceDriver FAIL: CAPTURED was " + captured
                    + " expected the calling instance " + instance);
            System.exit(5);
        }
        clearHook.invoke();
        System.out.println("DRIVER_PASS instance-method :redefine captures self via @This");
    }
}

File: src/test/clojure/nihilite/test/redefine_instance_driver.clj
(ns nihilite.test.redefine-instance-driver
  "Replaces src/test/java/nihilite/test/redefineInstanceDriver.java.

   Validates :redefine on an instance method: the bridge receives the
   calling instance via @This (verified by comparing identity).

   Invoked from build.clj as `java nihilite.test.redefineInstanceDriver`."
  (:require [nihilite.registry :as reg])
  (:import [net.bytebuddy.agent ByteBuddyAgent]
           [java.lang.instrument Instrumentation])
  (:gen-class
   :name nihilite.test.redefineInstanceDriver
   :prefix "rid-"
   :main true))

(def ^:private captured (atom nil))

(defn- generate-class-bytes! [options]
  (let [generate-class (Class/forName "clojure.core$generate_class")
        invoke-static (.getDeclaredMethod generate-class "invokeStatic"
                                         (into-array Class [Object]))]
    (.setAccessible invoke-static true)
    (let [[cname bytecode] (.invoke invoke-static nil (object-array [options]))]
      (clojure.lang.Compiler/writeClassFile cname bytecode)
      cname)))

(defn- generate-target-class! []
  (generate-class-bytes!
   {:name "nihilite.test.redefine_instance_driver.probe_target"
    :prefix "pt-"
    :impl-ns "nihilite.test.redefine-instance-driver"
    :main false
    :methods
    [[(with-meta (symbol "probe") {}) []
      (Class/forName "java.lang.String")]]}))

(defn gen-all!
  "Generates nihilite.test.redefine_instance_driver.probe_target (the
   target class whose probe() gets :redefine'd) into *compile-path*."
  []
  (generate-target-class!)
  nil)

;; Target stub body.

(defn pt-probe []
  "ORIGINAL-COUNTER=0")

(defn- fail! [why code]
  (println "redefineInstanceDriver FAIL:" why)
  (System/exit code))

(defn -main [& _args]
  (let [inst (ByteBuddyAgent/install)
        Agent (Class/forName "nihilite.kernel.Agent")
        agentmain (.getDeclaredMethod Agent "agentmain"
                                       (into-array Class [String Instrumentation]))]
    (.setAccessible agentmain true)
    (.invoke agentmain nil (object-array [nil inst]))
    ((resolve 'nihilite.kernel.installer/install) inst)


    (let [bridge (fn [self _args _method-name]
                   (reset! captured self)
                   "REDEFINED-AT-ARITY0")
          spec {:id "inst-redef"
                :target-internal "nihilite/test/redefine_instance_driver/probe_target"
                :method-name "probe"
                :position :redefine
                :arity 0
                :descriptor "()Ljava/lang/String;"
                :bridge bridge}]
      (reg/clear!)
      (reg/install! spec))

    (let [target (Class/forName "nihilite.test.redefine_instance_driver.probe_target")
          instance (.newInstance target (object-array []))
          result (.invoke (.getDeclaredMethod target "probe" (into-array Class []))
                          instance (object-array []))]
      (when (not= "REDEFINED-AT-ARITY0" result)
        (fail! (str "probe was \"" result "\" expected \"REDEFINED-AT-ARITY0\"") 3))
      (let [c @captured]
        (when (nil? c) (fail! "bridge did not run (CAPTURED null)" 4))
        (when (not (clojure.lang.Util/identical c instance))
          (fail! (str "CAPTURED was " c " expected " instance) 5))))
    (reg/clear!)
    (println "DRIVER_PASS instance-method :redefine captures self via @This")
    (System/exit 0)))

(when *compile-files*
  (gen-all!))
